### PR TITLE
Butchering proficiencies

### LINKED ISSUE #54392: Add new "butchery" proficiency
### Is your feature request related to a problem? Please describe.

We need a new butchery proficiency, its unrealistic when random dude with fine knife has same meat, bones and skin output as experienced butcher

### Solution you would like.

adding new proficiency "butchering", "expert butchering", "skinning" and "expert skinning" into the game (but not sure how it must be)

### Describe alternatives you have considered.

not doing it

add only one level of proficiency

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
Features "Butchering now require proficiency to do it fast and effectively"
#### Purpose of change
Close #54392
#### Describe the solution
Add four proficiencies: 
- Basic butchery (drastically reduces flesh obtained from butchering and speed of butchering), 
- Advanced butchery (same as basic, but has only little effect), 
- Basic and Advanced skinning (same as butchery)

Lack of butchery_basic applies little (10-15%) penalty to amount of meat, but huge (~2x) penalty to speed
Having butchery_basic still has some less than moderate speed penalty
Having butchery_adv removes this penalty

Lack of skinning_basic applies little penalty to speed, but massive penalty to skin dropped
Having skinning_basic applies less than moderate penalty to skin dropped
Having skinning_adv removes this penalty

#### Testing
~~TBD, trying to solve error in loops, that causes crazy things like @screenshot~~
UPD: current numbers are like this:
- for meat output (quick butchering cow with butcher knife):
- - unmodified game: 60 kg of meat
- - character with no proficiency: 53 kg of meat
- - character with basic proficiencies: 60 kg of meat
- - character with adv proficiency: 60 kg

- for time spend:
- - unmodified game:
image
- - character with no proficiency:
image
- - character with basic proficiencies (butchering, skinning):
image
- - character with adv proficiencies:
image
#### Additional context
I would like to add few more stuff, but it would not fit the game really - proper bleeding of the corpse, lack of bleeding causing it smell and taste bad are two what i thought about
Also maybe i should treat quick butchery more laxed than full butchery?

TODO in future prs: make butchery snippets, make event for butchery finish, make snippets for it also
maybe todo in future: bleeding or lack of it affect meat quality?

### CODE REVIEW

**Comment by NetSysFire** [comment]
Dont forget to add the butchering proficiency to some of the backgrounds/hobbies.

File: data/json/proficiencies/butchering.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    "time_to_learn": "4 h"
+  },
+  {
+    "type": "proficiency",
+    "id": "prof_skinning_adv",
+    "category": "prof_butchering",
+    "name": { "str": "Skinning Expert" },
+    "description": "You really know how to skin the animal.",
+    "can_learn": true,
+    "time_to_learn": "40 h"
```
Comment (TheShadowFerret) [inline data/json/proficiencies/butchering.json:36]:
```suggestion
    "time_to_learn": "40 h",
    "required_proficiencies": [ "prof_skinning_basic" ]
```

File: data/json/proficiencies/butchering.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    "time_to_learn": "4 h"
+  },
+  {
+    "type": "proficiency",
+    "id": "prof_butchering_adv",
+    "category": "prof_butchering",
+    "name": { "str": "Butchering Expert" },
+    "description": "There is very little meat left after your skillful knife movements.",
+    "can_learn": true,
+    "time_to_learn": "40 h"
```
Comment (TheShadowFerret) [inline data/json/proficiencies/butchering.json:18]:
```suggestion
    "time_to_learn": "40 h",
    "required_proficiencies": [ "prof_butchering_basic" ]
```

**Review by TheShadowFerret — COMMENTED** [review]
I think the X hunters should get either basic skinning or basic butchering, all of them *do* carry hunting knives.

File: data/json/hobbies.json [hunk 3]
```diff
@@ -836,7 +836,14 @@
     "points": -3,
     "traits": [ "ANTIFRUIT", "MEATARIAN" ],
     "skills": [ { "level": 2, "name": "survival" }, { "level": 2, "name": "cooking" } ],
-    "proficiencies": [ "prof_knives_familiar", "prof_knife_skills" ]
+    "proficiencies": [
+      "prof_knives_familiar",
+      "prof_knife_skills",
+      "prof_butchering_basic",
+      "prof_butchering_adv",
```
Comment (TheShadowFerret) [inline data/json/hobbies.json:843]: `amateur butcher`
```suggestion
```

Comment (GuardianDll) [reply data/json/hobbies.json:843]: I took the last place in reading competition

**Comment by harakka** [comment]
I'm not sure if evisceration distinct from butchery really makes sense. These things kinda go hand in hand in practice.

**Comment by GuardianDll** [comment]
sound reasonable, gotta throw it away same as bleeding prof

**Comment by Night-Pryanik** [comment]
I don't like the idea of corpse's meat magically disappearing into nowhere if you're not so skilled in butchering. Maybe the remaining meat from the corpse should be converted to meat scraps?

**Comment by GuardianDll** [comment]
it's not magically disappeared, it is transformed into butchery refuse

**Comment by Night-Pryanik** [comment]
> it's not magically disappeared, it is transformed into butchery refuse

Then I'm fine with that.

**Comment by IdleSol** [comment]
I don't think butchering an animal is such a complicated process that it requires several stages of training.

I propose the following option:
1. Basic animal butchery (basic skill, opens access to 2nd and 3rd item)
2. Skinning
3. Butchering animals for meat

Regarding internal organs, I don't really understand how this is different from normal dissection and why we need a separate profession. When we already have a lot of professions related to dissection. As well as a separate menu item.

I would rename dissecting people to dissection. (You know how to carefully dissect bodies and use the proper tools.) Add the appropriate anatomy and you get a complete extraction of internal organs.

**Standard character, no animal butchering skills.**
Takes longer, gets less. If you can do this, it blocks access to the butchering menu items. I'd replace them with a “sloppy butchering” item. So that there would be a distinction from quick butchering. What's the yield? Say no more than 60%. Or 20-60% depending on your learning progress.

Is it possible to train on zombies? Additionally, the general anatomy and anatomy of the animal (or non-animal) in question is studied.

**A character with learned basic animal butchering.**
Gains access to full (or without full butchering?) butchering menu. Makes faster, increases production to 100% (60-100 depending on learning progress).

Additional study of anatomy? Training on zombies (or zombie versions of animals)? Weaknesses?

**Comment by GuardianDll** [comment]
three skills do not have meaningful difference comparing to four in this pr

dissection is not touched in this pr at all

no, training weakpoints via butchery is no go

no, ui should not be changed

**Comment by IdleSol** [comment]
40 hours, really?

Learning the weaknesses in 1-2 hours. And to completely butcher a carcass, we learn 40 hours? Three hours per cow, that's roughly 13 cows. My character would sooner get a 10 in melee than eat all the meat he gets.

> three skills do not have meaningful difference comparing to four in this pr

Two for butchering, two for hides and one for organs. That makes five. Or have any of them already been canceled?

**Comment by GuardianDll** [comment]
\> Or have any of them already been canceled?

right above your first message is me removing offal prof

\> And to completely butcher a carcass, we learn 40 hours?

You learn 40 hours to clean off all the meat from the body, the last 15% , to just "get more than half" you learn 4 h

**Comment by HadeanLake** [comment]
50% meat loss for not having a proficience is a lot. How did you come up with this number? 
It takes 4h (7h if comparing no prof vs advanced) more already, does yield needs to be penalized so harshly as well?

**Comment by IdleSol** [comment]
1. Do you have no error in accounting for the `basic skinning`?
```
    if( corpse.harvest->has_entry_type( harvest_drop_skin ) && ( action == butcher_type::FULL ||
            action == butcher_type::QUICK || action == butcher_type::SKIN ) ) {
        time_to_cut *= 1.25 - ( 0.25 * you.get_proficiency_practice( proficiency_prof_skinning_adv ) );
    }
```
Where's the impact of the `Basic skinning`? Or do we study it, but it doesn't give anything in terms of speed?

2. I must have misunderstood something. The cutting time is calculated as:
```
        time_to_cut *= 1.25 - ( 0.25 * you.get_proficiency_practice( proficiency_prof_butchering_adv ) );
        time_to_cut *= 1.75 - ( 0.75 * you.get_proficiency_practice( proficiency_prof_butchering_basic ) );
...
        time_to_cut *= 1.25 - ( 0.25 * you.get_proficiency_practice( proficiency_prof_skinning_adv ) );
```
A quick butchery example. 

You know the advanced versions of skills: 27 minutes
You know the basic versions of skills: 27*1.25*1.25 = 42 minutes (vs 59 in the screenshot)
You don't know anything: 27*1.75*1.25 (?) = 59 minutes (vs 1 hour, and that doesn't display minutes)

**Comment by IdleSol** [comment]
2.7 times slower. We get 50% of the production. Or 0.5 / 2.7 =~ 18% efficiency. Isn't that a little harsh?

Alternative.
```
time_to_cut *= 0.7 * ( 1.75 - 0.25 * you.get_proficiency_practice( proficiency_prof_butchering_adv - 0.5 * you.get_proficiency_practice( proficiency_prof_butchering_basic) + 0.3 * ( 1.25 - 0.25 * you.get_proficiency_practice( proficiency_prof_skinning_basic)
```

Based on the fact that 30% of the body butchering time is taken up by skinning.

Has further potential for development. If instead of exact values we use a function to calculate proportions depending on the volume/mass of the animal. Conditionally for small animals 50/50, and for large animals 70/30.

**Comment by GuardianDll** [comment]
small animals do not have much larger amount of skin comparing to their weight, so i suggest to use formula from this PR: 40% for skin, 40% for butcher, 20% for rest activities

**Comment by HadeanLake** [comment]
> We talk about a person that lever butchered animal in their life, it is expected to mess up a lot of stuff and do it wrong

Can it be adjusted at least for small things, that are not as complex to butcher? Cant imagine someone (who is capable of making a salad) messing up preparing for cooking (in non-fancy way) non-fancy fish way so badly 50% of meat is lost.
istockphoto-2158177752-612x612

**Comment by IdleSol** [comment]
```
if( ( corpse.harvest->has_entry_type( harvest_drop_flesh ) ||
          corpse.harvest->has_entry_type( harvest_drop_offal ) ) && ( action == butcher_type::FULL ||
                  action == butcher_type::QUICK || action == butcher_type::FIELD_DRESS ) )
```
`action == butcher_type::FIELD_DRESS` Do I understand correctly that this is responsible for extracting internal organs? Should it be here?

Or do we think that butchering applies to both organs and meat?

**Comment by GuardianDll** [comment]
@IdleSol i still miss what you mean with your message

\> that are not as complex to butcher
is there even such thing as "not complex to butcher"? i assumed that even fish can be messed up pretty badly so you'd have only miserable pile of mush you can't even cook
maybe i need to revisit it

**Comment by IdleSol** [comment]
Maybe I'm confused myself. In the beginning, there was a profession of cutting + a profession of extraction of offal. The offal profession was abandoned. I thought that the time of organ extraction is always the same and does not depend on the study of the profession.

So I'm assuming that “action == butcher_type::FIELD_DRESS” is redundant.

But maybe you merged the two professions into one? And now the cutting profession includes both meat and offal extraction?

**Comment by GuardianDll** [comment]
that's correct, but i mostly asked about changes in butchery speed formula

**Comment by GuardianDll** [comment]
\> Can it be adjusted at least for small things, that are not as complex to butcher
For test i full butchered rabbit (only 35 minutes) and was able to get 600 g of meat (with all profs it's just 1 kg), so i guess it works fine even here

**Comment by Bone-blood-moss** [comment]
Food Preparation and Knife Skills should have Effect on Butchering skills. You know what to gather form the body but have not done it before. You're less likely to ruin the meat but you'll be slow.

**Comment by HadeanLake** [comment]
A lot of people mention how much time it takes them to butcher a deer they hunted (I summarized what I read as "about 9h might be spent on full butchering a deer for the first time and only 2h after years of experience"). But it is much harder to find any info about how much they wasted. As if it is only a matter of how many steaks (instead of meat that is only good for meatballs) they got.

Hopefully we will get a meat expert to clarify and correct formulas one day.

**Comment by IdleSol** [comment]
about this?
```
time_to_cut *= 0.4 * K1 + 0.4 * K2 + 0.2
```

If so, then.

As per your comment, I divide the body butchering time into three parts:
1. 40% of the time, I spend on skinning. (Basic and Advanced skinning)
2. 40% of the time, I spend harvesting the meat and offal. (Basic and Advanced butchery)
3. 20% of the time, I spend on everything else. 

In total, that's 100% of the time I spend butchering the body of an animal. Assuming I've learned all the skills.

If I don't know how to skin an animal, I will spend more time. But not the total time will increase, but only the part of it that relates to skinning.

That is, without skills it will be 40% multiplied by the K2 factor (from the post earlier). You use 1.25 if there is no basic profession and 1 if there is. So the total cutting time will be:
`40% * K2 + 40 % + 20%` 

By analogy. Butchering for meat, we get a coefficient K1 = {1, 1.25, 1.75} depending on knowledge. Or K1_alt = {1, 1.25, 1.75 *1.25} is how you have it in the script now.

Total total time would be `40% * K1 + 40% * K2 + 20%`

**Comment by GuardianDll** [comment]
I see what you mean now, I'll tweak it tomorrow to apply speed penalty relatively to the type of butchery (and how long skinning/butchering step lasts in it)

**Comment by RedMisao** [comment]
I think the mass penalties are too harsh, considering there's also time penalties. I do think the latter should be the primary component over the former, given that unless you somehow _destroy_ or discard otherwise usable flesh, it can still be handled by meat scraps.

Speaking from experience butchering small animals (chicken, fish, rabbit), there's only two possibilities:
1. You either do it efficiently and obtain the maximum possible amount of meat (optimal amount), or 
2. You do it inneficiently by wasting time while getting less than the optimal amount, which is usually stick to the bones or to the skin by cutting it improperly.

There is no way you lose ~40% (25 kg) of the maximum amount of meat (60 kg) by spending twice the total time handling the carcass. That means there's almost half still stuck to the bones or the skin, which is not just useless but also can't be scrapped away. One thing is to hack so bad you mix the cuts (i.e. drumstick with a chunk of the thighs), another is to make it **100% useless** for consumption.

**Comment by IdleSol** [comment]
For example, you could damage the intestines. And its contents will spill out onto the meat. I think that's a typical rookie mistake. Is this meat considered spoiled? Does it meet the 50% mark? I don't know.

Forgetting to drain the blood or not draining all the blood. Does this sort of count as spoiled meat? Would it be considered spoiled under cataclysmic conditions?

**Comment by GuardianDll** [comment]
No, lack of drain is, at first, not emulated, and second, do not spoil the meat quality to the unusable level
Damaging digestive system, spilling organ content, ineffective cuts (tho this one is tricky, because we don't have "meat on a bone" thing, only them being separately) is what i considered the consequences of bad skill.
@RedMisao i should remind that we talk about quick butchery, not full one, it is lossy butchery type on its own
Despite i don't like it (because i already spend time making it), it is a strong argument to make butchery not affect amount of meat obtained, but affect only the speed of it

So, in short:
Lack of butchery_basic applies little (10-15%) penalty to amount of meat, but huge (same as now) penalty to speed
Having butchery_basic still has some less than moderate speed penalty
Having butchery_adv removes this penalty

Lack of skinning_basic applies little penalty to speed, but massive penalty to skin dropped
Having skinning_basic applies less than moderate penalty to skin dropped
Having skinning_qdv removes this penalty

And the speed of butchery should be revisited in the future to match real life numbers, not 2 hours to butcher an entire cow

Sounds good for me

**Comment by IdleSol** [comment]
Is it possible that I'm misunderstanding? Is there a mistake?
```
 // 40% of butchering and gutting, 40% of skinning, 20% another activities
        time_to_cut *= 1 + ( penalty_small * ( 1 - butch_adv ) * 0.4 );
        time_to_cut *= 1 + ( penalty_big * ( 1 - butch_basic ) * 0.4 );
        time_to_cut *= 1 + ( penalty_small * ( 1 - skin_basic ) * 0.4 );
```
This turns into a sequential multiplication of penalties, doesn't it? I.e. again we take the total cutting time and multiply it by three penalties. Instead of taking a part of the total time and penalizing only that part (with the appropriate penalty).

Or have you decided not to do that?

And there's a strange way the 40% factor is taken into account. We don't take 40% of the time, we take 40% of the penalty.

**Comment by GuardianDll** [comment]
that's on me
would this one work now?

File: src/activity_handlers.cpp [hunk 4]
```diff
@@ -1057,6 +1101,22 @@ static bool butchery_drops_harvest( item *corpse_item, const mtype &mt, Characte
             roll = 0;
         }
 
+        const double butch_basic = you.get_proficiency_practice( proficiency_prof_butchering_basic );
+        const double butch_adv = you.get_proficiency_practice( proficiency_prof_butchering_basic );
```
Comment (HadeanLake) [inline src/activity_handlers.cpp:1105]: buid fails with `unused variable 'butch_adv' [-Werror,-Wunused-variable]`
also, 
```patch
-const double butch_adv = you.get_proficiency_practice( proficiency_prof_butchering_basic );
+const double butch_adv = you.get_proficiency_practice( proficiency_prof_butchering_adv );
```

Comment (GuardianDll) [reply src/activity_handlers.cpp:1105]: 😮‍💨 
```suggestion
```

**Comment by HadeanLake** [comment]
pr body shows that full butchering takes twice as much time with 0 prof
penalties to achieve that with new formula are:
 penalty_small = 0.5;
 penalty_big = 1.5;

**Comment by IdleSol** [comment]
@GuardianDll

If I have not made a mistake in my calculations, then everything is correct. Although such a record, in my opinion, is difficult to understand.

**Comment by GuardianDll** [comment]
what record?

**Comment by IdleSol** [comment]
| Prof | 0.25/0.75| 05/1.5 |
|--------|--------|--------|
| none | x 1.5 | x 2.0 |
| b_butch | x 1.4 | x 1.8 |
| b_skin | x 1.4 | x 1.8 |
| b_butch & b_skin | x 1.3 | x 1.6 |
| b_butch & a_butch| x 1.1 | x 1.2 |
| b_butch & a_butch & b_skin | x 1.0 | x 1.0 |

**Comment by IdleSol** [comment]
```
    double butch_basic = you.get_proficiency_practice( proficiency_prof_butchering_basic );
    double butch_adv = you.get_proficiency_practice( proficiency_prof_butchering_adv );
    double skin_basic = you.get_proficiency_practice( proficiency_prof_skinning_basic );
    double penalty_small = 0.25;
    double penalty_big = 0.75;
--    int prof_penalty = 0; 
++  int prof_butch_penalty = penalty_big * ( 1 - butch_basic ) + penalty_small * ( 1 - butch_adv );
++  int prof_skin_penalty = penalty_small * ( 1 - skin_basic );

    if( action == butcher_type::FULL ) {
        // 40% of butchering and gutting, 40% of skinning, 20% another activities
--        prof_penalty += ( time_to_cut * 0.4 ) * penalty_small * ( 1 - butch_adv );
--        prof_penalty += ( time_to_cut * 0.4 ) * penalty_big * ( 1 - butch_basic );
--        prof_penalty += ( time_to_cut * 0.4 ) * penalty_small * ( 1 - skin_basic );

--        time_to_cut += prof_penalty;

++      time_to_cut *= 0.4 * ( 1 + prof_butch_penalty ) + 0.4 * ( 1 + prof_skin_penalty ) + 0.2;
    }

    if( action == butcher_type::QUICK ) {
        // 70% of butchery, 15% skinning, 15% another activities
--        prof_penalty += ( time_to_cut * 0.7 ) * penalty_small * ( 1 - butch_adv );
--        prof_penalty += ( time_to_cut * 0.7 ) * penalty_big * ( 1 - butch_basic );
--        prof_penalty += ( time_to_cut * 0.15 ) * penalty_small * ( 1 - skin_basic );

--        time_to_cut += prof_penalty;

++      time_to_cut *= 0.7 * ( 1 + prof_butch_penalty ) + 0.15 * ( 1 + prof_skin_penalty ) + 0.15;
    }

    if( action == butcher_type::FIELD_DRESS ) {
--        prof_penalty += time_to_cut * penalty_small * ( 1 - butch_adv );
--        prof_penalty += time_to_cut * penalty_big * ( 1 - butch_basic );

--        time_to_cut += prof_penalty;

++      time_to_cut *=  1 + prof_butch_penalty
    }

    if( action == butcher_type::SKIN ) {
--        prof_penalty += time_to_cut  * penalty_small * ( 1 - skin_basic );

--        time_to_cut += prof_penalty;
++      time_to_cut *=  1 + prof_skin_penalty
    }
```

**Comment by IdleSol** [comment]
By the way, what if we do this:

```
double knife_skills = you.get_proficiency_practice( proficiency_prof_knife_skills );
double penalty_big = 1.25 + 0.25 * (1 - knife_skills);
```

**Comment by GuardianDll** [comment]
i like it, but i suspect having no knife skill above of not knowing how to butcher need to penalize thing even more (but also train knife skill)
```c++
penalty_big = 1.5 + 0.5 * (1 - knife_skills)
```
?

**Comment by GuardianDll** [comment]
oh, wrong, proficiency specifically says about chopping, not at carving meat from the bones and skin from the meat

**Comment by IdleSol** [comment]
Why not? Although I thought the 1.5 factor already incorporated poor knife handling.

>oh, wrong, proficiency specifically says about chopping, not at carving meat from the bones and skin from the meat, so it doesn't fit here

We have a strange way of butchering bodies in our game. For some reason we immediately separate meat from bones and cut it into equal pieces. From this point of view, this skill could work.

File: src/activity_handlers.cpp [hunk 5]
```diff
@@ -811,6 +815,37 @@ int butcher_time_to_cut( Character &you, const item &corpse_item, const butcher_
     if( corpse_item.has_flag( flag_QUARTERED ) ) {
         time_to_cut /= 4;
     }
+
+    double butch_basic = you.get_proficiency_practice( proficiency_prof_butchering_basic );
+    double butch_adv = you.get_proficiency_practice( proficiency_prof_butchering_adv );
+    double skin_basic = you.get_proficiency_practice( proficiency_prof_skinning_basic );
+    double penalty_small = 0.5;
+    double penalty_big = 1.5;
+    int prof_penalty = 0;
```
Comment (GuardianDll) [inline src/activity_handlers.cpp:824]:
```suggestion
```
